crates/composer/src/sound.rs

```diff
@@ -0,0 +1,176 @@
+            |err| eprintln!("Got cpal stream error callback: {err}."),
+            None,
+        )?;
+
+        Ok(Self { source_tx, play_delay, too_early_plays, _stream })
+    }
+
+    pub(crate) fn play<S: Source<Item = f32> + Send + 'static>(&self, source: S) {
+        // TODO(Matej): use timestamp from the event itself once we have it.
+        let play_at_timestamp = current_timestamp() + self.play_delay;
+
+        // TODO(Matej): we are in fact double-boxing because DynamicMixerController internally adds
+        // another box. But we need a sized type to send it through threads. We could make this
+        // method non-generic, but that would be less flexible, so just accept it for now.
+        let source = Box::new(source);
+
+        self.source_tx
+            .send(TimedSource { source, play_at_timestamp })
+            .expect("source receiver should be still alive");
+    }
+
+    /// Get "too early plays" counter since the last call of this method.
+    pub(crate) fn fetch_too_early_plays(&self) -> u64 {
+        self.too_early_plays.swap(0, Ordering::SeqCst)
+    }
+}
+
+/// An f32 [rodio::source::Source] with UNIX timestamp of desired play time attached.
+struct TimedSource {
+    source: Box<dyn Source<Item = f32> + Send + 'static>,
+    play_at_timestamp: Duration,
+}
+
+/// A sort of manual implementation of the closure used as cpal audio data callback, for tidiness.
+struct AudioCallback {
+    mixer_controller: Arc<DynamicMixerController<f32>>,
+    mixer: DynamicMixer<f32>,
+    source_rx: Receiver<TimedSource>,
+    too_early_plays: Arc<AtomicU64>,
+    stream_start: Option<StreamStart>,
+}
+
+impl AudioCallback {
+    fn new(
+        mixer_controller: Arc<DynamicMixerController<f32>>,
+        mixer: DynamicMixer<f32>,
+        source_rx: Receiver<TimedSource>,
+        too_early_plays: &Arc<AtomicU64>,
+    ) -> Self {
+        let too_early_plays = Arc::clone(too_early_plays);
+        Self { mixer_controller, mixer, source_rx, too_early_plays, stream_start: None }
+    }
+
+    fn fill_data(&mut self, data_out: &mut [f32], info: &OutputCallbackInfo) {
+        let stream_timestamp = info.timestamp();
+        let stream_start =
+            self.stream_start.get_or_insert_with(|| StreamStart::new(stream_timestamp));
+
+        // At least on Linux ALSA, cpal gives very strange stream timestamp on very first call.
+        if stream_start.callback > stream_timestamp.callback {
+            eprintln!("cpal's stream timestamp jumped backwards, resetting stream start.");
+            *stream_start = StreamStart::new(stream_timestamp);
+        }
+
+        // UNIX timestamp of when the buffer filled during this call will be actually played.
+        let buffer_playback_timestamp = stream_start.realtime
+            + (stream_timestamp.playback.duration_since(&stream_start.callback).expect(
+                "current playback timestamp should be larger than start callback timestamp",
+            ));
```

Comment:
I had hardtime understanding this timstamp-jutsu. But after some sketches in a notebook it clicked and I was then able to simplify it a bit in https://github.com/tonarino/acoustic_profiler/pull/39

Reply:
hehe, I don't think it will be comprehensible though


![IMG_7712](https://user-images.githubusercontent.com/16712262/235147747-3063503b-1f57-4107-a93e-c1ff009d026a.JPG)


the "ball" is the sound we're trying to position, the line on right is the destination buffer with the question mark being the delay we need to compute. On the left there are the two independent epochs. The arrows are... well I guess just my thoughts as I was figuring it out .D